### PR TITLE
EventStream

### LINKED ISSUE #299: API: Refactoring API Eventstream
- With API Handler struct, Design `Streamer` or `EventStream` struct to work with `http.Flusher`
- related to: #286

### PR DESCRIPTION
**Pull request by anarcher**
### Github Issue

Related to #299 

### Background

It is a refactoring `streaming` func

### Solution

I think we can use it like below:

```
event := fmt.Sprintf("address-%s", address)
es := NewDefaultEventStream(w, r)
es.Render(blk)
es.Run(observer.BlockAccountObserver, event)
```

- If you have an implementation of `common.Serializable` interface, You don't need to add `es.On(onFunc)`.  A `OnSerializableFunc` handle it by default.  it's optional. 
 

### Possible Drawbacks

- For testing, I have to do to `EventStream.Start` returns `func()`.  If we need other needs, we will change the `func` to `struct`.
- No source comment. If this PR is approved, I will add comments about it. 
- For preventing data race condition, we should use  `go-observable@dev` branch.  It has already contained our `Gopkg.lock`. :-)

### CODE REVIEW

File: lib/api_stream.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+func (s *EventStream) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	s.Run(w, r)
+}
+
+func (s *EventStream) Run(w http.ResponseWriter, r *http.Request) {
+	s.Start(w, r)()
+}
+
+func (s *EventStream) Start(w http.ResponseWriter, r *http.Request) func() {
```
[the remark this answers is not in the document]
Comment (anarcher) [reply lib/api_stream.go]: It a good question, The`start` can be better than `Start`. But If some pkg want to use `Start` and then some works (e.g. logging), I think `Start` can be better than `start`.

File: lib/api_stream.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+func (s *EventStream) SetContentType(ct string) {
+	s.ContentType = ct
+}
+
+func (s *EventStream) Before(beforeFunc func()) {
+	s.beforeFunc = beforeFunc
+}
+
+func (s *EventStream) On(onFunc OnFunc) {
```
[the remark this answers is not in the document]
Comment (anarcher) [reply lib/api_stream.go]: I will think that :-)  Firstly, I think the constructor with `onFunc` too.   But If we use the `OnSeriailizeFunc` by default.  I felt that  `NewEventStream()` might support required arguments. 
I think it is still an open question for me.  or using dynamic args? `(ob *obserser,event string,onfunc ...OnFunc)` ?

[the remark this answers is not in the document]
Comment (kfangw) [reply lib/api_stream.go]: I mean
```NewEventStream(ob,OnSerializableFunc,"test1","test2")```
I prefer explicit behavior :)

File: lib/api_stream.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kfangw) [reply lib/api_stream.go]: I couldn't think that. That might happen.
If then, how about changing the function name to `Register`?
`Start` seems like the streaming is started when the function called.

File: lib/api_stream.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+		select {
+		case msg <- payload:
+		case <-stop:
+			return
+		}
+	}
+	s.observer.On(event, onFunc)
+
+	if s.beforeFunc != nil {
+		go s.beforeFunc()
```
[the remark this answers is not in the document]
Comment (kfangw) [reply lib/api_stream.go]: Or... we can just remove this feature.
What I intended in the previous was returning existing data thru streaming
If we set the specs of the streaming like that only recent data is returned, we can remove the feature.
How do you think about this?

[the remark this answers is not in the document]
Comment (Geod24) [reply lib/api_stream.go]: I'd be in favor of @kfangw 's solution

File: lib/api_stream.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (anarcher) [reply lib/api_stream.go]: I think `RenderFunc` is a better name then the name `OnFunc` :-)

File: lib/api_stream_test.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+)
+
+func TestAPIStreamRun(t *testing.T) {
+	tests := []struct {
+		name       string
+		makeStream func(*observable.Observable) *EventStream
+		trigger    func(*observable.Observable)
+		respFunc   func(testing.TB, *http.Response)
+	}{
+		{
```
Comment (Geod24) [inline lib/api_stream_test.go]: While the double block ? It's rather hard to read

File: lib/api_stream.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (anarcher) [reply lib/api_stream.go]: `EventStream` doesn't handle `beforeFunc` is OK for me :-) 
Or, What do you think about `beforeFunc doesn't use trigger`?  It means using `EventStream.RenderFunc` (`OnFunc`). It makes smaller handler codes.. Is it right?  :-)

File: lib/api_stream.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+	ContentType string
+	observer    *observable.Observable
+	beforeFunc  func()
+	onFunc      OnFunc
+}
+
+type OnFunc func(args ...interface{}) ([]byte, error)
+
+var DefaultContentType = "application/json"
+var OnSerializableFunc = func(args ...interface{}) ([]byte, error) {
```
Comment (Geod24) [inline lib/api_stream.go]: `DefaultContentType` can be `const`. Also indentation is off.

File: lib/api_stream.go [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+	s, ok := args[1].(common.Serializable)
+	if !ok {
+		return nil, fmt.Errorf("this is not serializable") // TODO(anarcher): Error type
```
Comment (Geod24) [inline lib/api_stream.go]: Could you add a test case for this ?

File: lib/api_stream.go [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+			fn = OnSerializableFunc
+		}
+
+		var (
+			payload []byte
+			err     error
+		)
+
+		if len(args) > 1 {
+			payload, err = fn(args...)
```
Comment (Geod24) [inline lib/api_stream.go]: This does not seem to be covered by the tests ?

File: lib/api_stream.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Geod24) [reply lib/api_stream.go]: I understood as "`EventStream` doesn't handle `beforeFunc`"

File: lib/api_stream.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (anarcher) [reply lib/api_stream.go]: How about the name `Prepare` or `Init`?  `run := es.Prepare();run()` v ` run := es.Start() ; run() `

File: lib/api_stream.go (hunk 6 above)
Comment (anarcher) [reply lib/api_stream.go]: What kind of test case?  I will add more test code. :-)

File: lib/api_stream.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (anarcher) [reply lib/api_stream.go]: @kfangw , @Geod24 You agreed that "ES doesn't handle the `beforeFunc`"?  :-)

File: lib/api_stream_test.go (hunk 4 above)
Comment (anarcher) [reply lib/api_stream_test.go]: IMHO, To write is harder to read.. hahaha ㅠㅠ

File: lib/api_stream.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (anarcher) [reply lib/api_stream.go]: `PrepareRun`

**Comment by anarcher** [comment]
I redesign the EventStream like below:

```
es := NewDefaultEventStream(w, r)
es.Render(blk)
es.Run(observer.BlockAccountObserver, event)
```

File: lib/api_stream_test.go [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+				var ba block.BlockAccount
+				require.Nil(t, json.Unmarshal(s.Bytes(), &ba))
+				require.Nil(t, s.Err())
+				require.Equal(t, ba, *block.NewBlockAccount("hello", 100, "tx1-tx1"))
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
```
Comment (Charleslee522) [inline lib/api_stream_test.go]: 👍

**Pull request merged**